Source/ACE/Network/GameMessages/Messages/GameMessageFellowshipFullUpdate.cs

```diff
@@ -0,0 +1,110 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using ACE.Common.Extensions;
+using ACE.Entity.Actions;
+using ACE.Entity;
+
+namespace ACE.Network.GameMessages.Messages
+{
+    public class GameMessageFellowshipFullUpdate : GameMessage
+    {
+        public GameMessageFellowshipFullUpdate(Session session)
+            : base(GameMessageOpcode.GameEvent, GameMessageGroup.Group09)
+        {
+            Fellowship fellowship = session.Player.Fellowship;
+            ActionChain chain = new ActionChain();
+            chain.AddAction(session.Player, () =>
+            {
+                Writer.Write(session.Player.Guid.Full);
+                Writer.Write(session.GameEventSequence++);
+                Writer.Write((uint)GameEvent.GameEventType.FellowshipFullUpdate);
+
+                // the current number of fellowship members
+                Writer.Write((UInt16)fellowship.FellowshipMembers.Count);
+
+                // ????
+                Writer.Write((byte)0x10);
+                Writer.Write((byte)0x00);
+
+                // --- FellowInfo ---
+                
+                foreach (Player fellow in fellowship.FellowshipMembers)
+                {
+                    if (fellow.Guid.Full != session.Player.Guid.Full)
+                    {
+                        chain.AddAction(fellow, () => { WriteFellow(fellow); });
```

Comment:
The ActionChain is a method of delaying action (not locking), it stops two threads from accessing the data at the same time by delaying work until its sure only one element will access that data.

This means it will delay the work until the lock is released, and you wont have to worry about deadlocks.

We didn't use a locking scheme for synchronizing between different interacting game entities due to lock-cycles and deadlocking dangers.